pkg/buildscript/queries.go

```diff
@@ -169,15 +194,45 @@ func (b *BuildScript) getSolveNode() (*Value, error) {
 
 		return nil
 	}
+
 	if node := search(b.raw.Assignments); node != nil {
 		return node, nil
 	}
+	return nil, errNodeNotFound
+}
+
+func (b *BuildScript) getSolveNode(target string) (*Value, error) {
+	node, err := b.getTargetNode(target)
+	if err != nil {
+		return nil, errs.Wrap(err, "Could not get target node")
+	}
+
+	// If the target is the solve function, we're done.
+	if isSolveFuncName(node.FuncCall.Name) {
+		return node, nil
+	}
+
+	// Otherwise, the "src" key contains a reference to the solve node. Look over the build expression
+	// again for that referenced node.
+	for _, arg := range node.FuncCall.Arguments {
+		if arg.Assignment == nil {
+			continue
+		}
+		a := arg.Assignment
+		if a.Key == srcKey && a.Value.Ident != nil {
+			node, err := b.getSolveNode(*a.Value.Ident)
+			if err != nil {
+				return nil, errs.Wrap(err, "Could not get solve node from target")
+			}
+			return node, nil
+		}
+	}
```

Comment:
Why can't this solve node be identified by the `getTargetNode` above? Is it because the `target` value wasn't found? Can you please add a more detailed comment.

Reply:
Consider this build script snippet:

```
runtime = state_tool_artifacts_v1(src = sources)
sources = solve(...)
main = runtime
```

`getTargetNode()` would return `sources` (`main -> runtime -> src`). Since `sources` is not a solve node (it's a reference), we have to look again for the `sources` node. The second time, `getTargetNode("sources")` will give us the solve node we need.